pow/src/main/java/tech/pegasys/teku/pow/Eth1BlockFetcher.java

```diff
@@ -0,0 +1,153 @@
+  private final NavigableSet<UnsignedLong> blocksToRequest = new TreeSet<>();
+  private boolean requestInProgress = false;
+  private boolean active = false;
+
+  public Eth1BlockFetcher(
+      final Eth1EventsChannel eth1EventsChannel,
+      final Eth1Provider eth1Provider,
+      final TimeProvider timeProvider,
+      final UnsignedLong cacheDuration) {
+    this.eth1EventsChannel = eth1EventsChannel;
+    this.eth1Provider = eth1Provider;
+    this.timeProvider = timeProvider;
+    this.cacheDuration = cacheDuration;
+  }
+
+  public synchronized void onInSync(final UnsignedLong latestCanonicalBlockNumber) {
+    if (active) {
+      return;
+    }
+    active = true;
+    LOG.debug("Beginning back-fill of Eth1 blocks");
+    backfillEth1Blocks(latestCanonicalBlockNumber);
+  }
+
+  public void fetch(final BigInteger fromBlock, final BigInteger toBlock) {
+    synchronized (this) {
+      if (!active) {
+        // No point starting to request data if we haven't yet reached the start of the chain
+        // We'll just wind up downloading a heap of blocks we then throw away
+        return;
+      }
+      for (BigInteger block = fromBlock;
+          block.compareTo(toBlock) <= 0;
+          block = block.add(BigInteger.ONE)) {
+        blocksToRequest.add(UnsignedLong.valueOf(block));
+      }
+    }
+    requestNextBlockIfRequired();
+  }
+
+  private void requestNextBlockIfRequired() {
+    UnsignedLong blockToRequest;
+    synchronized (this) {
+      if (requestInProgress || blocksToRequest.isEmpty()) {
+        return;
+      }
+      blockToRequest = blocksToRequest.last();
+      blocksToRequest.remove(blockToRequest);
+      requestInProgress = true;
+    }
+    requestBlock(blockToRequest)
+        .always(
+            () -> {
+              synchronized (Eth1BlockFetcher.this) {
+                requestInProgress = false;
+              }
+              requestNextBlockIfRequired();
+            });
+  }
+
+  private SafeFuture<Void> requestBlock(final UnsignedLong blockNumberToRequest) {
+    // Note: Not using guaranteed requests here - if the Eth1 chain is temporarily unavailable
+    // we may miss some blocks but that's better than potentially getting stuck retrying a block
+    LOG.debug("Requesting block {}", blockNumberToRequest);
+    return eth1Provider
+        .getEth1BlockFuture(blockNumberToRequest)
+        .thenAccept(
+            block -> {
+              if (isAboveLowerBound(UnsignedLong.valueOf(block.getTimestamp()))) {
+                postBlock(block);
+              } else {
+                // Every block before the one we just fetched must be outside of the range
+                synchronized (Eth1BlockFetcher.this) {
+                  // All blocks at or before this number must be before the cache period
+                  blocksToRequest.headSet(blockNumberToRequest, true).clear();
+                }
+              }
+              requestNextBlockIfRequired();
```

Comment:
This line seems redundant. Every time `requestNextBlockIfRequired ` calls `requestBlock`, it sets `requestInProgress` to `true`. However, `requestNextBlockIfRequired` does not set `requestInProgress` to false. So `requestNextBlockIfRequired` should just return early every time its called inside `requestBlock`.

Reply:
Yes it is, that logic got moved to the `.always` but I missed this call.